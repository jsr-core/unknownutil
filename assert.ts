import type { Predicate } from "./type.ts";

/**
 * A factory function that generates assertion error messages.
 */
export type AssertMessageFactory = (
  x: unknown,
  pred: Predicate<unknown>,
  name?: string,
) => string;

/**
 * The default factory function used to generate assertion error messages.
 */
export const defaultAssertMessageFactory: AssertMessageFactory = (
  x,
  pred,
  name,
) => {
  const p = pred.name || "anonymous predicate";
  const t = typeof x;
  const v = JSON.stringify(x, null, 2);
  return `Expected ${
    name ?? "a value"
  } that satisfies the predicate ${p}, got ${t}: ${v}`;
};

let assertMessageFactory = defaultAssertMessageFactory;

/**
 * Represents an error that occurs when an assertion fails.
 */
export class AssertError extends Error {
  /**
   * Constructs a new instance.
   * @param message The error message.
   */
  constructor(message?: string) {
    super(message);

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, AssertError);
    }

    this.name = this.constructor.name;
  }
}

/**
 * Sets the factory function used to generate assertion error messages.
 *
 * ```ts
 * import { is, setAssertMessageFactory } from "@core/unknownutil";
 *
 * setAssertMessageFactory((x, pred) => {
 *   if (pred === is.String) {
 *     return `Expected a string, got ${typeof x}`;
 *   } else if (pred === is.Number) {
 *     return `Expected a number, got ${typeof x}`;
 *   } else if (pred === is.Boolean) {
 *     return `Expected a boolean, got ${typeof x}`;
 *   } else {
 *     return `Expected a value that satisfies the predicate, got ${typeof x}`;
 *   }
 * });
 * ```
 *
 * @param factory The factory function.
 */
export function setAssertMessageFactory(factory: AssertMessageFactory): void {
  assertMessageFactory = factory;
}

/**
 * Asserts that the given value satisfies the provided predicate.
 *
 * It throws {@linkcode AssertError} if the value does not satisfy the predicate.
 *
 * ```ts
 * import { assert, is } from "@core/unknownutil";
 *
 * const a: unknown = "hello";
 * assert(a, is.String);
 * // a is now narrowed to string
 * ```
 *
 * @param x The value to be asserted.
 * @param pred The predicate function to test the value against.
 * @param options Optional configuration for the assertion.
 * @returns The function has a return type of `asserts x is T` to help TypeScript narrow down the type of `x` after the assertion.
 */
export function assert<T>(
  x: unknown,
  pred: Predicate<T>,
  options: { message?: string; name?: string } = {},
): asserts x is T {
  if (!pred(x)) {
    throw new AssertError(
      options.message ?? assertMessageFactory(x, pred, options.name),
    );
  }
}
