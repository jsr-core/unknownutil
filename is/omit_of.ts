import type { FlatType } from "../_typeutil.ts";
import type { WithPredObj } from "../_annotation.ts";
import type { Predicate } from "../type.ts";
import { isObjectOf } from "./object_of.ts";

/**
 * Return a type predicate function that returns `true` if the type of `x` is `Omit<ObjectOf<T>, K>`.
 *
 * To enhance performance, users are advised to cache the return value of this function and mitigate the creation cost.
 *
 * ```typescript
 * import { as, is } from "@core/unknownutil";
 *
 * const isMyType = is.OmitOf(is.ObjectOf({
 *   a: is.Number,
 *   b: is.String,
 *   c: as.Optional(is.Boolean),
 * }), ["a", "c"]);
 * const a: unknown = { a: 0, b: "a" };
 * if (isMyType(a)) {
 *   const _: { b: string } = a;
 * }
 * ```
 */
export function isOmitOf<
  T extends Record<PropertyKey, unknown>,
  P extends Record<PropertyKey, Predicate<unknown>>,
  K extends keyof T,
>(
  pred: Predicate<T> & WithPredObj<P>,
  keys: K[],
):
  & Predicate<FlatType<Omit<T, K>>>
  & WithPredObj<P> {
  const s = new Set(keys);
  const predObj = Object.fromEntries(
    Object.entries(pred.predObj).filter(([k]) => !s.has(k as K)),
  );
  return isObjectOf(predObj) as
    & Predicate<FlatType<Omit<T, K>>>
    & WithPredObj<P>;
}
