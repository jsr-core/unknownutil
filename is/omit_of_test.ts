import { assertEquals } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { assertType, type IsExact } from "@std/testing/types";
import { is } from "./mod.ts";
import { isOmitOf } from "./omit_of.ts";

Deno.test("isOmitOf<T, K>", async (t) => {
  const pred = is.ObjectOf({
    a: is.Number,
    b: is.String,
    c: is.Boolean,
  });

  await t.step("returns properly named predicate function", async (t) => {
    assertEquals(typeof isOmitOf(pred, ["b"]), "function");
    await assertSnapshot(t, isOmitOf(pred, ["b"]).name);
    await assertSnapshot(t, isOmitOf(isOmitOf(pred, ["b"]), ["c"]).name);
  });

  await t.step("returns true on Omit<T, K> object", () => {
    assertEquals(
      isOmitOf(pred, ["b"])({ a: 0, b: undefined, c: true }),
      true,
    );
    assertEquals(isOmitOf(pred, ["b", "c"])({ a: 0 }), true);
  });

  await t.step("returns false on non Omit<T, K> object", () => {
    assertEquals(
      isOmitOf(pred, ["b"])("a"),
      false,
      "Value is not an object",
    );
    assertEquals(
      isOmitOf(pred, ["b"])({ a: 0, b: "a", c: "" }),
      false,
      "Object have a different type property",
    );
  });

  await t.step("predicated type is correct", () => {
    const a: unknown = { a: 0, b: "a", c: true };

    if (isOmitOf(pred, ["b"])(a)) {
      assertType<
        IsExact<typeof a, { a: number; c: boolean }>
      >(true);
    }

    if (isOmitOf(isOmitOf(pred, ["b"]), ["c"])(a)) {
      assertType<
        IsExact<typeof a, { a: number }>
      >(true);
    }
  });

  await t.step("with symbol properties", async (t) => {
    const b = Symbol("b");
    const c = Symbol("c");
    const pred = is.ObjectOf({
      a: is.Number,
      [b]: is.String,
      [c]: is.Boolean,
    });

    await t.step("returns properly named predicate function", async (t) => {
      assertEquals(typeof isOmitOf(pred, [b]), "function");
      await assertSnapshot(t, isOmitOf(pred, [b]).name);
      await assertSnapshot(t, isOmitOf(isOmitOf(pred, [b]), [c]).name);
    });

    await t.step("returns true on Omit<T, K> object", () => {
      assertEquals(
        isOmitOf(pred, [b])({ a: 0, [b]: undefined, [c]: true }),
        true,
      );
      assertEquals(isOmitOf(pred, [b, c])({ a: 0 }), true);
    });

    await t.step("returns false on non Omit<T, K> object", () => {
      assertEquals(
        isOmitOf(pred, [b])("a"),
        false,
        "Value is not an object",
      );
      assertEquals(
        isOmitOf(pred, [b])({ a: 0, [b]: "a", [c]: "" }),
        false,
        "Object have a different type property",
      );
    });

    await t.step("predicated type is correct", () => {
      const x: unknown = { a: 0, [b]: "a", [c]: true };

      if (isOmitOf(pred, [b])(x)) {
        assertType<
          IsExact<typeof x, { a: number; [c]: boolean }>
        >(true);
      }

      if (isOmitOf(isOmitOf(pred, [b]), [c])(x)) {
        assertType<
          IsExact<typeof x, { a: number }>
        >(true);
      }
    });
  });
});
