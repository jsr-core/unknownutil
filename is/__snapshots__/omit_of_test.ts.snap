export const snapshot = {};

snapshot[`isOmitOf<T, K> > returns properly named predicate function 1`] = `
"isObjectOf({
  a: isNumber,
  c: isBoolean
})"
`;

snapshot[`isOmitOf<T, K> > returns properly named predicate function 2`] = `"isObjectOf({a: isNumber})"`;

snapshot[`isOmitOf<T, K> > with symbol properties > returns properly named predicate function 1`] = `
"isObjectOf({
  a: isNumber,
  Symbol(c): isBoolean
})"
`;

snapshot[`isOmitOf<T, K> > with symbol properties > returns properly named predicate function 2`] = `"isObjectOf({a: isNumber})"`;
